### ISSUE TITLE
RUF013 does not check usage of implicit None in class constructors

### ISSUE THREAD

**Issue by mikeweltevrede** [astral-sh/ruff#12403]
I noticed that rule [RUF013](https://docs.astral.sh/ruff/rules/implicit-optional/) did not check usage of implicit optionals in the class constructors `__init__()`. I would expect to see `RUF013 PEP 484 prohibits implicit 'Optional'`, but instead, it passes the checks.

# List of keywords you searched for before creating this issue
- RUF013
- Optional

# A minimal code snippet that reproduces the bug
```python
class MyClass:
    def __init__(self, a: int = None): ...
```

# The command you invoked
- `ruff check --select RUF013`
- `pre-commit run --all-files`

# The current Ruff settings
```toml
[tool.ruff]
line-length = 120
target-version = "py39"

[tool.ruff.lint]
extend-select = [
    ...,
    "RUF",
    ...,
]
```

# The current Ruff version
`0.5.3` but also tried with earlier versions `0.3.0` and `0.4.0`.

**Comment by charliermarsh** [comment]
This actually is flagged for me in the playground: https://play.ruff.rs/ed0f6c85-acdf-4e74-a3de-c17deb79f0ef

**Comment by charliermarsh** [comment]
And on the command-line too:

```
❯ ruff check foo.py --select RUF --isolated
foo.py:2:27: RUF013 PEP 484 prohibits implicit `Optional`
  |
1 | class MyClass:
2 |     def __init__(self, a: int = None): ...
  |                           ^^^ RUF013
  |
  = help: Convert to `Optional[T]`
```

Maybe something else going on? Is that the exact MRE you used?

**Comment by mikeweltevrede** [comment]
Thanks @charliermarsh, I will go through our situation with a fine-tooth comb and see if something else might be interfering with this. If it works well in the ruff playground for you, then it must be something on our side that I am not seeing.

**Comment by charliermarsh** [comment]
We have some logic in there to detect annotations that implicitly allow `None` (like `Any`). If you find a concrete example, I might be able to map it back to that logic (and figure out whether it's intended or not).

**Comment by mikeweltevrede** [comment]
Hi @charliermarsh, I tried to reproduce again but could not. Must have been some other setting in our project that got updated. Thanks for your quick response.
